compiler/rustc_middle/src/ty/abstract_const.rs

```diff
@@ -127,68 +29,55 @@ TrivialTypeTraversalAndLiftImpls! {
     NotConstEvaluatable,
 }
 
+pub type BoundAbstractConst<'tcx> = Result<Option<EarlyBinder<ty::Const<'tcx>>>, ErrorGuaranteed>;
+
 impl<'tcx> TyCtxt<'tcx> {
+    /// Returns a const with substs applied by
+    pub fn bound_abstract_const(
+        self,
+        uv: ty::WithOptConstParam<DefId>,
+    ) -> BoundAbstractConst<'tcx> {
+        self.thir_abstract_const_opt_const_arg(uv).map(|ac| ac.map(|ac| EarlyBinder(ac)))
+    }
     #[inline]
     pub fn thir_abstract_const_opt_const_arg(
         self,
         def: ty::WithOptConstParam<DefId>,
-    ) -> Result<Option<&'tcx [Node<'tcx>]>, ErrorGuaranteed> {
+    ) -> Result<Option<ty::Const<'tcx>>, ErrorGuaranteed> {
         if let Some((did, param_did)) = def.as_const_arg() {
             self.thir_abstract_const_of_const_arg((did, param_did))
         } else {
             self.thir_abstract_const(def.did)
         }
     }
-}
 
-#[instrument(skip(tcx, f), level = "debug")]
-pub fn walk_abstract_const<'tcx, R, F>(
-    tcx: TyCtxt<'tcx>,
-    ct: AbstractConst<'tcx>,
-    mut f: F,
-) -> ControlFlow<R>
-where
-    F: FnMut(AbstractConst<'tcx>) -> ControlFlow<R>,
-{
-    #[instrument(skip(tcx, f), level = "debug")]
-    fn recurse<'tcx, R>(
-        tcx: TyCtxt<'tcx>,
-        ct: AbstractConst<'tcx>,
-        f: &mut dyn FnMut(AbstractConst<'tcx>) -> ControlFlow<R>,
-    ) -> ControlFlow<R> {
-        f(ct)?;
-        let root = ct.root(tcx);
-        debug!(?root);
-        match root {
-            Node::Leaf(_) => ControlFlow::CONTINUE,
-            Node::Binop(_, l, r) => {
-                recurse(tcx, ct.subtree(l), f)?;
-                recurse(tcx, ct.subtree(r), f)
+    pub fn expand_bound_abstract_const(
+        self,
+        ct: BoundAbstractConst<'tcx>,
+        substs: &[GenericArg<'tcx>],
+    ) -> Result<Option<Const<'tcx>>, ErrorGuaranteed> {
+        let ac = ct?;
+        let Some(ac) = ac else {
+            return Ok(None);
+        };
+
+        let ac = ac.subst(self, substs);
+        struct Expander<'tcx> {
+            tcx: TyCtxt<'tcx>,
+        }
+        impl<'tcx> TypeFolder<'tcx> for Expander<'tcx> {
```

Comment:
`FallibleTypeFolder`

I think having `tcx.expand_constants` method: `fn expand_abstract_consts<T: TypeFoldable<'tcx>>(self, value: T) -> Result<Const<'tcx>, ErrorGuaranteed>` is probably a bit nicer. If `bound_abstract_const` returns `Ok(None)` then you can keep the `Unevaluated` const